pages/api/internal/assets/[anchor]/[assetType]/index.js

```diff
@@ -140,17 +146,31 @@ export default async function handle(req, res) {
 
 		if (existingAsset) {
 			await connectWithExistingAsset(existingAsset, nft)
-
+			rmSync(uploadedFile.path)
 			return res.json({
 				existingAsset
 			})
 		} else {
 			try {
-				writeFileSync(path.join(STORAGE_DIR, fileHash), fileContent)
+				// Split files per Smart contract. 
+				// This is a logical separation and also reduces the risk of Filesystem-overload
+				// due to too many files in the same directory
+				const storageFileName = path.join(nft.contract.csn, fileHash)
+				const targetFile = path.join(STORAGE_DIR, storageFileName)
+				const targetDir =path.dirname(targetFile)
+
+				// Create contract-directory in case it does not exist
+				if(!existsSync(targetDir)) {
+					mkdirSync(targetDir) 
+				}
```

Comment:
This if+mkdir has a TOCTOU issue, and will throw an exception if you're unlucky with parallel uploads. Never check whether a directory exists and then create it if not. Instead, always attempt to create it and catch the kernel's EEXISTS response. The kernel has an atomic view on the filesystem.

Reply:
will use the `recursive` approach, which corresponds to `mkdir -p`, which AFAIK solves for this issue
https://nodejs.org/api/fs.html#fsmkdirsyncpath-options

Reply:
Done in the course of https://github.com/authenticvision/digitalsoul/pull/41/commits/8807190ea224f95c7a949f5f541f4b70aea8b306